cpp/src/jit/parser.cpp

```diff
@@ -387,52 +326,34 @@ std::string ptx_parser::parse()
 
   input_arg_list.clear();
   // Go directly to the .func mark
-  const size_t length = no_comments.size();
-  size_t start        = no_comments.find(".func") + 5;
-  if (start == length + 5) {
+  size_t offset = no_comments.find(".func");
+  if (offset == no_comments.size()) {
     printf("No function (.func) found in the input ptx code.\n");
     exit(1);
   }
-  size_t stop = start;
-  while (stop < length && no_comments[stop] != '{') { stop++; }
 
-  std::string function_header = std::string(no_comments, start, stop - start);
+  auto f = no_comments.cbegin() + offset + 5;  // 5 = length of ".func"
+  auto l = std::find(f, no_comments.cend(), '{');
 
-  stop++;
-  start = stop;
-
-  int bra_count = 0;
-  while (stop < length) {
-    if (no_comments[stop] == '{') bra_count++;
-    if (no_comments[stop] == '}') {
-      if (bra_count == 0) {
-        break;
-      } else {
-        bra_count--;
-      }
+  auto f2 = std::next(l);
+  auto l2 = std::find_if(f2, no_comments.cend(), [brace_count = 0](auto c) mutable {
+    if (c == '{') ++brace_count;
+    if (c == '}') {
+      if (brace_count == 0) return true;  // find matching } to first found {
+      --brace_count;
     }
-    stop++;
-  }
-
-  std::vector<std::string> function_body_output =
-    parse_function_body(std::string(no_comments, start, stop - start));
-
-  std::string function_header_output = parse_function_header(function_header);
-
-  std::string final_output = function_header_output + "\n";
-  final_output += " asm volatile (\"{\");";
-  for (int i = 0; i < function_body_output.size(); i++) {
-    if (function_body_output[i].find("ret;") != std::string::npos) {
-      final_output += "  asm volatile (\"bra RETTGT;\");\n";
-      continue;
-    }
-    final_output += "  " + function_body_output[i] + "\n";
-  }
-  final_output += " asm volatile (\"RETTGT:}\");";
-
-  final_output += "}";
-
-  return final_output;
+    return false;
+  });
+
+  auto const fn_body_output   = parse_function_body(std::string(f2, l2));  // TODO wtf - order?
```

Comment:
What was this comment referring to? 
```suggestion
  auto const fn_body_output   = parse_function_body(std::string(f2, l2));
```

Reply:
Ah -- yes still need to fix this. It refers this [commit](https://github.com/rapidsai/cudf/pull/5603/commits/2d816f5e0a559a2e14c36a9521953fcefd02a542) which took me several hours to try down. The order matters.

Reply:
Oh yikes... I wasn't expecting these methods to have side effects. I thought they were free functions!

Reply:
Comment updated.